sylvia-derive/src/querier.rs

```diff
@@ -0,0 +1,119 @@
+use convert_case::{Case, Casing};
+use proc_macro2::{Ident, TokenStream};
+use proc_macro_error::emit_error;
+use quote::quote;
+use syn::parse::{Parse, Parser};
+use syn::spanned::Spanned;
+use syn::{GenericParam, ImplItem, ItemImpl, ItemTrait, TraitItem};
+
+use crate::check_generics::CheckGenerics;
+use crate::crate_module;
+use crate::message::{MsgField, MsgVariant};
+use crate::parser::{MsgAttr, MsgType};
+
+pub struct Querier<'a> {
+    variants: Vec<MsgVariant<'a>>,
+}
+
+impl<'a> Querier<'a> {
+    pub fn for_contract(source: &'a ItemImpl, generics: &[&'a GenericParam]) -> Self {
+        let mut generics_checker = CheckGenerics::new(generics);
+        let variants: Vec<_> = source
+            .items
+            .iter()
+            .filter_map(|item| match item {
+                ImplItem::Method(method) => {
+                    let msg_attr = method.attrs.iter().find(|attr| attr.path.is_ident("msg"))?;
+                    let attr = match MsgAttr::parse.parse2(msg_attr.tokens.clone()) {
+                        Ok(attr) => attr,
+                        Err(err) => {
+                            emit_error!(method.span(), err);
+                            return None;
+                        }
+                    };
+
+                    if attr == MsgType::Query {
+                        Some(MsgVariant::new(&method.sig, &mut generics_checker, attr))
+                    } else {
+                        None
+                    }
+                }
+                _ => None,
+            })
+            .collect();
+        Self { variants }
+    }
+
+    pub fn for_interface(source: &'a ItemTrait, generics: &[&'a GenericParam]) -> Self {
```

Comment:
`for_contract` and `for_interface` are almost identical - at first glance, they only differ on the `source` type, but then just reach very similar `items` under the hood, doing exactly the same things on them.

That means that you want to abstract type differences. Here you basically reach the `Method`  variant, very similar for both, only to access common fields `sig` and `attrs` which are identical at this point.

What you do to make it properly is either:
1. Implement extension trait on `TraitItemMethod` and `ImplItemMethod`, which gives you access to `attrs` and `sig`, and then extension trait for `ItemTrait` and `ImplTrait` to give you back iterator over something implementing your trait - it involves generic traits and possibly GATs, it is a difficult approach, most possibly overkill.
2. Map `ItemTrait` and `ItemImpl` to some iterator over temporary struct before, and pass this iterator into this function argument. This looks like the best approach. Ideally you create a trait on both types with a function returning iterator over your simplified `items` containing only signatures and attributes of whatever methods you found.